beancount/plugins/close_tree.py

```diff
@@ -0,0 +1,98 @@
+"""This plugin inserts close directives for all of an account's descendants when an account is
+closed. Unopened parent accounts can also be closed. Any explicitly specified close is left
+untouched.
+
+For example, given this:
+
+```
+2017-11-10 open Assets:Brokerage:AAPL
+2017-11-10 open Assets:Brokerage:ORNG
+2018-11-10 close Assets:Brokerage  ; this account does not necessarily need to be opened
+```
+
+the plugin turns it into:
+```
+2017-11-10 open Assets:Brokerage:AAPL
+2017-11-10 open Assets:Brokerage:ORNG
+2018-11-10 close Assets:Brokerage:AAPL
+2018-11-10 close Assets:Brokerage:ORNG
+```
+
+Invoke this plugin _after_ any plugins that generate `open` directives for account trees that you
+want to auto close. An example is the `auto_accounts` plugin that ships with Beancount:
+
+```
+plugin "beancount.plugins.auto_accounts"
+plugin "beancount.plugins.close_tree"
+```
+"""
+
+import time
+from beancount.core import data
+from beancount.core.data import Open, Close
+
+DEBUG = 0
+__plugins__ = ('close_tree',)
+
+
+def close_tree(entries, unused_options_map):
+    """Insert close entries for all subaccounts of a closed account.
+
+    Args:
+      entries: A list of directives. We're interested only in the Open/Close instances.
+      unused_options_map: A parser options dict.
+    Returns:
+      A tuple of entries and errors. """
+
+    start_time = time.time()
+    close_count = 0
+    new_entries = []
+    errors = []
+
+    opens = [e.account for e in entries if isinstance(e, Open)]
+    closes = [e.account for e in entries if isinstance(e, Close)]
+
+    for entry in entries:
+        if isinstance(entry,  Close):
+            subaccounts = [a for a in opens if entry.account + ':' in a and a not in closes]
```

Comment:
Use a set for speed

Reply:
Speed is not a concern, so leaving this as such (technically, I believe a list might be slightly faster in this case as the hashing of sets is avoided, and we don't do the expensive lookups or removals).

Reply:
Changed.